### PR TITLE
Add LoadingSkeleton component

### PR DESCRIPTION
**Pull request by michael-yx-wu**
#### Checklist

- [ ] Include tests
- [x] Update documentation

#### Changes proposed in this pull request:

Add a `LoadingSkeleton` component to be reused in other components (such as Blueprint Table loading states). The API will generally look something like this

```ts
<LoadingSkeleton isLoading={true}>
  {renderSomeChildren()}
</LoadingSkeleton>
```

and will help us reduce branching code paths later on. Shoutout to @themadcreator for the initial suggestion for this component!

#### Reviewers should focus on:

- The CSS API seems pretty limited and I'm having a hard time imagining why anyone would be forced to use it instead of the Javascript API. Should we keep it (as in, should we document it)?
- Is the Javascript API expressive enough? Do we want to include support for intent?

### CODE REVIEW

**Comment by michael-yx-wu** [comment]
Disabled tests for so docs can be previewed, will add tests in the morning!

**Comment by giladgray** [comment]
I don't think this solves the problem correctly. The idea behind a loading skeleton is to hint at the content being loaded, but every bone here looks the same. This is what I'd expect to see for the example on the docs:

skeleton

Notice how the "bone" for the heading is much larger than for the paragraph, which echoes what the real content looks like. This is how Facebook's skeleton works. I imagine this could be done quite simply by cloning the children and adding a class that hides text and sets gray background. Then the bone would adapt to the styles of where it's used.

**Comment by llorca** [comment]
@giladgray sounds pretty sweet indeed, that's the ideal scenario. In your example, the block of text contains 4 lines, how do you create 4 bars that have the same width as every line in the block of text?

File: packages/core/src/common/classes.ts [hunk 1]
```diff
@@ -19,6 +19,7 @@ export const ALIGN_LEFT = "pt-align-left";
 export const ALIGN_RIGHT = "pt-align-right";
 export const INLINE = "pt-inline";
 export const FILL = "pt-fill";
+export const ANIMATED = "pt-animated";
```
Comment (llorca) [inline packages/core/src/common/classes.ts]: Nice, would you mind alphabetically ordering this list of modifiers? We do the same below for components

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss [hunk 2]
```diff
@@ -0,0 +1,127 @@
+// Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+// Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+// of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+// and https://github.com/palantir/blueprint/blob/master/PATENTS
+
+@import "../../common/variables";
+
+.pt-loading-skeleton-bones {
```
Comment (llorca) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: `-bones` or `-container`?

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  display: flex;
+  flex-wrap: wrap;
+  align-items: center;
+  width: 100%;
+  height: 100%;
+}
+
+@mixin bone-margin($margin) {
+  .pt-loading-skeleton-bone-#{$margin} {
+    margin-right: percentage($margin / 100);
```
Comment (llorca) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: Conceptually this is great, but I feel weird about `margin-right`. Why not a `width` between 70% and 100%?

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  &.pt-animated {
+    // negative animation delay ensures animations are synced
+    animation: 1.75s linear -7s infinite normal none running glow;
+  }
+}
+
+.pt-loading-skeleton-content {
+  opacity: 0;
+  animation: ($pt-transition-duration * 5) fade-in;
+  animation-fill-mode: forwards;
```
Comment (llorca) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: Can we collapse this line with the `animation` rule above?

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+.pt-loading-skeleton-bone {
+  border-radius: 2px;
+  background-color: rgba($gray1, 0.2);
+  width: 100%;
+  height: $pt-grid-size * 0.6;
+
+  &.pt-animated {
+    // negative animation delay ensures animations are synced
+    animation: 1.75s linear -7s infinite normal none running glow;
```
Comment (llorca) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: This line seems quite verbose for a simple glow animation. Why not `glow ($pt-transition-duration * 20) infinite`?

File: packages/core/src/components/loading-skeleton/loadingSkeleton.tsx [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    randomWidth?: boolean;
+}
+
+export interface ILoadingSkeletonState {
+    animated?: boolean;
+    rightMargins?: number[];
+}
+
+export class LoadingSkeleton extends AbstractComponent<ILoadingSkeletonProps, ILoadingSkeletonState> {
```
Comment (llorca) [inline packages/core/src/components/loading-skeleton/loadingSkeleton.tsx]: Would prefer calling it `Skeleton`, simply. Loading doesn't add much, it's also a term we use as a modifier. We've done a good job so far at keeping our component names short and sweet. Similarly to `Spinner` (which isn't called `LoadingSpinner`), I believe this component should live under the "Progress" section of the docs

File: packages/core/examples/loadingSkeletonExample.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                <Switch
+                    checked={animated}
+                    key="animated"
+                    label="Animated"
+                    onChange={this.handleAnimatedChange}
+                />,
+                <Switch
+                    checked={isLoading}
+                    key="loading"
+                    label="Show Loading Skeleton"
```
Comment (llorca) [inline packages/core/examples/loadingSkeletonExample.tsx]: "**S**entence **c**asing" instead of "**T**itle **C**asing"

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss (hunk 4 above)
Comment (llorca) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: Also, wouldn't a simpler `transition` rule work here? Why is an animation needed?

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+/*
+Loading Skeleton
+
+`LoadingSkeleton`s allow you to conditionally show a loading animation in places where you have
+content that fits into rectangular containers.
+
+<div class="pt-callout pt-intent-warning pt-icon-warning-sign">
+  <h5>Note</h5>
+  <p>
```
Comment (llorca) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: This `<p>` element is unnecessary, can remove it 👍

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+Styleguide components.loading-skeleton.css
+*/
+
+/*
+JavaScript API
+
+The `LoadingSkeleton` component is available in the __@blueprintjs/core__ package.
+Make sure to review the [general usage docs for JS components](#components.usage).
+
+Unlike the CSS API, the React Component is able to guarantee that animations remain in sync as bones
```
Comment (llorca) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: "component"

File: packages/core/src/components/loading-skeleton/loadingSkeleton.tsx [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+            this.setState({ rightMargins });
+        } else if (nextNumBones > numBones) {
+            if (rightMargins.length !== nextNumBones) {
+                for (let i = numBones; i < nextNumBones; i++) {
+                    rightMargins.push(this.generateRightMargin(randomWidth));
+                }
+            }
+
+            // sync animations
+            this.setState({ rightMargins });
```
Comment (llorca) [inline packages/core/src/components/loading-skeleton/loadingSkeleton.tsx]: Is it possible to sync the animations via CSS only? If all the bones appear at the same time, maybe?

File: packages/core/src/components/loading-skeleton/loadingSkeleton.tsx [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+                this.setState({ animated: false }, () => setTimeout(this.setState({ animated: true }), 250));
+            }
+        } else if (nextNumBones < numBones) {
+            this.boneRefs.splice(nextNumBones, numBones - nextNumBones);
+        }
+    }
+
+    public render() {
+        const className = classNames(Classes.LOADING_SKELETON, {
+            "pt-loading-skeleton-bones": this.props.isLoading,
```
Comment (llorca) [inline packages/core/src/components/loading-skeleton/loadingSkeleton.tsx]: Should use the `Classes` object

File: packages/core/src/components/loading-skeleton/loadingSkeleton.tsx [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+export interface ILoadingSkeletonState {
+    animated?: boolean;
+    rightMargins?: number[];
+}
+
+export class LoadingSkeleton extends AbstractComponent<ILoadingSkeletonProps, ILoadingSkeletonState> {
+    public static defaultProps: ILoadingSkeletonProps = {
+        animated: true,
```
Comment (llorca) [inline packages/core/src/components/loading-skeleton/loadingSkeleton.tsx]: I think we should remove both this prop and the corresponding CSS modifier. We don't have a use case for non-animated skeletons and I don't foresee any at the moment

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss (hunk 2 above)
Comment (michael-yx-wu) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: is `container` more consistent across BP?

File: packages/core/src/common/classes.ts (hunk 1 above)
Comment (michael-yx-wu) [inline packages/core/src/common/classes.ts]: ah, yes!

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss (hunk 2 above)
Comment (llorca) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: I would say so. I like the skeleton/bone metaphor but IMO it's pushed a little too far with "bones"

**Comment by michael-yx-wu** [comment]
@giladgray How do we interpolate that the paragraph is going to render as 4 rows of text? In the above card, we have

```
<LoadingSkeleton>
  <h5>Loading Skeleton</h5>
  <p>/* text here */</p>
</LoadingSkeleton>
```

I think there are 2 problems with the adding a modifier approach:

- if the child is a react component, we can't easily make it render with a class unless the child handles the `className` prop similar to BP components (i'd love to be wrong about this)
- in the case of child components that can take multiple rows (as in the case of the `<p>` child), how do we know how many bars to use without first rendering the content?

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss (hunk 5 above)
Comment (llorca) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: `normal`, `none` and `running` are all defaults. What about `animation: ($pt-transition-duration * 20) linear ($pt-transition-duration * -70) glow;`?

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss (hunk 9 above)
Comment (llorca) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: Bones can't be added or removed via CSS only, so I wonder if it's worth mentioning

File: packages/docs/src/styles/_examples.scss [hunk 13]
```diff
@@ -213,6 +213,12 @@ $options-margin: $pt-grid-size * 3;
   animation: docs-wiggle-rotate $pt-transition-duration $pt-transition-ease infinite;
 }
 
+.docs-loading-skeleton-example-box {
+  flex: 1;
+  margin-right: $pt-grid-size * 1.5;
+  height: $pt-grid-size * 15;
```
Comment (llorca) [inline packages/docs/src/styles/_examples.scss]: This line adds unnecessary extra space at the bottom of the card:
image

File: packages/core/examples/loadingSkeletonExample.tsx [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    public renderExample() {
+        const className = classNames("pt-card", "docs-loading-skeleton-example-box", {
+            "pt-loading": this.state.isLoading,
+        });
+
+        return (
+            <div className={className}>
+                <LoadingSkeleton {...this.state}>
+                    <h5><a href="#">Loading Skeleton</a></h5>
+                    <p>
```
Comment (llorca) [inline packages/core/examples/loadingSkeletonExample.tsx]: Let's remove this `<p>`, it adds unnecessary `margin-bottom` to the block of text:
image

File: packages/docs/src/styles/_examples.scss (hunk 13 above)
Comment (michael-yx-wu) [inline packages/docs/src/styles/_examples.scss]: if the height isn't fixed, the card size will change as it transitions between the loading / non-loading states

Comment (llorca) [inline packages/docs/src/styles/_examples.scss]: Hmm, so is this how any skeleton container should be used? What's wrong with the container size changing? Seems fine to me, in a real app it will only adjust once since the skeleton is only shown initially

Comment (michael-yx-wu) [inline packages/docs/src/styles/_examples.scss]: the bones vertically justify to the available space. if there is no inherent height in the containing parent, then the bones will be squashed together. of course we could add some vertical margins to make sure they are spaced at least X pixels from each other?

Comment (llorca) [inline packages/docs/src/styles/_examples.scss]: Yup that sounds good. Try `$pt-grid-size` for the margin

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss (hunk 4 above)
Comment (michael-yx-wu) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: @llorca i think if we try to collapse those two lines, we're forced to fill in more values to take advantage of the `animation` shorthand, but let me try again

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss (hunk 2 above)
Comment (michael-yx-wu) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: okay `container` it is. you have @amingione to thank for the the metaphor :)

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss (hunk 3 above)
Comment (michael-yx-wu) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: sure, i'm going to try width instead

File: packages/core/src/components/loading-skeleton/_loading-skeleton.scss (hunk 5 above)
Comment (michael-yx-wu) [inline packages/core/src/components/loading-skeleton/_loading-skeleton.scss]: i tried 2 seconds and it felt slow, but i'll change it and you can let me know how you feel.

File: packages/core/src/components/loading-skeleton/loadingSkeleton.tsx (hunk 10 above)
Comment (michael-yx-wu) [inline packages/core/src/components/loading-skeleton/loadingSkeleton.tsx]: this is handling for a rather specific case where you're adding more bones after the component has already mounted. the new bones start animating without regard to the old bone animations.

**Comment by giladgray** [comment]
I have many concerns about the API presented in this PR and I don't think we'll be able to rectify them through code review alone. I think our best path forward will be to take this component back to the drawing board and really **nail down the API before we build anything**. That's not to say we'll throw away this code, we just need to think through usage patterns and the appropriate API before actually building the React component. 💅 

@michael-yx-wu how do you feel about working with @llorca and me early next year to build this out the right way? 🛠

File: packages/core/src/components/loading-skeleton/loadingSkeleton.tsx (hunk 11 above)
Comment (michael-yx-wu) [inline packages/core/src/components/loading-skeleton/loadingSkeleton.tsx]: the `classNames` function expects a string literal here

Comment (llorca) [inline packages/core/src/components/loading-skeleton/loadingSkeleton.tsx]: ES6 tho? `[Classes.LOADING_SKELETON_BONES]: this.props.isLoading,`

**Comment by michael-yx-wu** [comment]
@giladgray Could you explain which parts of the API are troubling you? As commented above, I don't think we can 1:1 match the shape of the underlying content with a single `Skeleton` component  because we often don't know what that content looks like until we try to render it. You could, however, accomplish this by using multiple `Skeleton` components together and using some custom styles. I feel like this isn't too bad for things like sidebars, conditional form fields, etc. the structure of your app would likely force you to use individual `Skeleton`s anyway.

Would it be better if users could additionally use the API to describe the shape of the data so `Skeleton` doesn't have to do any guess work?

**Comment by michael-yx-wu** [comment]
If you feel strongly about the API needing serious redesign, let me know when would be a good time in January and I'd be happy to fly out then!

**Comment by giladgray** [comment]
To me, the point of `Skeleton` is that I already know what my component looks like, how it's laid out: there's a heading up top, then a nice image, then some small text, then some footer buttons (for instance). But I'm still loading the data, so I'd like to show placeholders for _each of these elements that give a rough estimate of their eventual shape and size_. The current API has _no way_ of mocking out a component like this, it just gives you X random bones that only vary in length. 

Blueprint components should do as much as possible for the consumers out of the box. The current API is still too far from that philosophy. @michael-yx-wu we're _super down_ to collaborate with you on this after break, and we'd be more than happy to host you in the design studio to jam in person! 🎸

**Comment by giladgray** [comment]
we'll revisit this component in early january
